### PR TITLE
Add ignoreFunctions: [] to unit-allowed-list

### LINKED ISSUE #5851: Add ignoreFunctions: [] to unit-allowed-list
### What steps are needed to reproduce the bug?

Use css/scss functions that include a disallowed unit character.

```scss
button {
  border: 1px solid hsl(20deg 0% 20% / 80%);
  background-color: lighten(green, 20%);
}

```

### What Stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "unit-allowed-list": "px"
  }
}

```

### How did you run Stylelint?

CLI `stylelint **/*.scss`

### Which version of Stylelint are you using?

14.2.0

### What did you expect to happen?

Units within function to be ignored.

### What actually happened?

The following problems were reported

```
2:25 errorUnexpected unit "deg" (unit-allowed-list)
2:31 errorUnexpected unit "%" (unit-allowed-list)
2:34 errorUnexpected unit "%" (unit-allowed-list)
2:40 errorUnexpected unit "%" (unit-allowed-list)
3:36 errorUnexpected unit "%" (unit-allowed-list)
```

### Does the bug relate to non-standard syntax?

Yes, SCSS color functions and vanilla HSL functions

### Proposal to fix the bug

Add `ignore: ["inside-function"]` secondary option to unit-allowed-list, similar fix to https://github.com/stylelint/stylelint/issues/5140

### PR DESCRIPTION
**Pull request by treemmett**
> Which issue, if any, is this issue related to?

Closes #5851 

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/rules/unit-allowed-list/__tests__/index.js [hunk 1]
```diff
@@ -340,3 +347,34 @@ testRule({
 		},
 	],
 });
+
+testRule({
+	ruleName,
+
+	config: [['px', 'em'], { ignoreFunctions: ['hsl', '/^rgb/'] }],
```
Comment (ybiquitous) [inline lib/rules/unit-allowed-list/__tests__/index.js:354]: [suggest] What do you think about adding a test case for a raw `RegExp` object?

E.g.

```suggestion
	config: [['px', 'em'], { ignoreFunctions: ['hsl', '/^rgb/', /-gradient$/] }],
```

**Review by ybiquitous — COMMENTED** [review]
@treemmett Thank you for creating the pull request! It looks almost good to me. 👍🏼

File: lib/rules/unit-allowed-list/__tests__/index.js (hunk 1 above)
Comment (treemmett) [reply lib/rules/unit-allowed-list/__tests__/index.js]: @ybiquitous regex test has been added

**Review by ybiquitous — APPROVED** [review]
Thank you! LGTM 👍🏼

File: lib/rules/unit-allowed-list/README.md [hunk 2]
```diff
@@ -113,3 +113,40 @@ a { -moz-border-radius-topright: 20rem; }
 ```css
 a { height: 100%; }
 ```
+
+### `ignoreFunctions: ["/regex/", /regex/, "string"]`
+
+Ignore units that are inside a function.
```
Comment (jeddy3) [inline lib/rules/unit-allowed-list/README.md:119]:
```suggestion
Ignore units that are inside of the specified functions.
```

Let's match the format above.

File: lib/rules/unit-allowed-list/README.md [hunk 3]
```diff
@@ -113,3 +113,40 @@ a { -moz-border-radius-topright: 20rem; }
...
+```css
+a {
+  background-image: linear-gradient(hsla(162deg, 51%, 35%, 0.8), hsla(62deg, 51%, 35%, 0.8));
+}
+```
+
+<!-- prettier-ignore -->
+```css
+a {
+  width: width: calc(100% - 10px);
```
Comment (jeddy3) [inline lib/rules/unit-allowed-list/README.md:150]:
```suggestion
  width: calc(100% - 10px);
```

File: lib/rules/unit-allowed-list/index.js [hunk 4]
```diff
@@ -48,9 +50,19 @@ function rule(listInput, options) {
...
+					// Ignore wrong units within `url` function
+					if (valueLowerCase === 'url') {
+						return false;
+					}
+
+					const ignoreFunctions = (options && options.ignoreFunctions) || [];
+
+					if (ignoreFunctions.length && matchesStringOrRegExp(valueLowerCase, ignoreFunctions)) {
+						return false;
+					}
```
Comment (jeddy3) [inline lib/rules/unit-allowed-list/index.js:65]: Let's use the `optionMatches` util here:

```js
if (optionsMatches(options, 'ignoreFunctions', valueLowerCase)) return false;
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for the pull request.

I've requested a couple of changes. Looks good otherwise.

**Comment by treemmett** [comment]
@jeddy3 Changes are added

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (jeddy3) [inline .vscode/settings.json:1]: Let's not include this file in the PR.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for making the changes.

One more request, and I think we're ready to merge.

**Review by jeddy3 — APPROVED** [review]
Thanks, LGTM!

**Comment by jeddy3** [comment]
> - Added: `ignoreFunctions: []` to `unit-allowed-list` ([#5857](https://github.com/stylelint/stylelint/pull/5857)).

**Pull request merged**